source/extensions/common/async_files/async_file_manager_factory.cc

```diff
@@ -0,0 +1,72 @@
+#include "source/extensions/common/async_files/async_file_manager_factory.h"
+
+#include <memory>
+#include <string>
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/protobuf/utility.h"
+#include "source/extensions/common/async_files/async_file_manager_thread_pool.h"
+
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace AsyncFiles {
+
+SINGLETON_MANAGER_REGISTRATION(async_file_manager_factory_singleton);
+
+class AsyncFileManagerFactoryImpl : public AsyncFileManagerFactory {
+public:
+  static std::shared_ptr<AsyncFileManagerFactory> singleton(Singleton::Manager* singleton_manager);
+  std::shared_ptr<AsyncFileManager> getAsyncFileManager(
+      const envoy::extensions::common::async_files::v3::AsyncFileManagerConfig& config,
+      Api::OsSysCalls* substitute_posix_file_operations = nullptr)
+      ABSL_LOCKS_EXCLUDED(mu_) override;
+
+private:
+  absl::Mutex mu_;
+  absl::flat_hash_map<std::string, std::shared_ptr<AsyncFileManager>>
+      managers_ ABSL_GUARDED_BY(mu_);
+  absl::flat_hash_map<std::string,
+                      envoy::extensions::common::async_files::v3::AsyncFileManagerConfig>
+      configs_ ABSL_GUARDED_BY(mu_);
```

Comment:
nit: just have a single map with a struct containing both the manager and config?

Reply:
That was very lazy of me, sorry. Done.